Add forward, backward and background color to badge designer

SegiNyn opened the pull request.

ThiefMaster commented on indico/modules/designer/client/js/index.js:
```diff
@@ -424,8 +430,63 @@ import {$T} from 'indico/utils/i18n';
+        }
+      }
+    }
+    const randomZIndexValue = Math.floor(Math.random() * 10) + 5;
```
Why is this needed? Using randomness seems really strange here.

SegiNyn replied:
I can change this to always increase/decrease by 1. it would work also

ThiefMaster commented on indico/modules/events/registration/badges.py:
```diff
@@ -83,7 +83,8 @@ def _draw_badge(self, canvas, registration, template, tpl_data, pos_x, pos_y):
         # Print images first
         image_placeholders = {name for name, placeholder in placeholders.items() if placeholder.is_image}
-        items = sorted(tpl_data.items, key=lambda item: item['type'] not in image_placeholders)
+        items = sorted(tpl_data.items, key=lambda item: (int(item['zIndex']) if 'zIndex' in item else 65535,
```
You can use `.get()` here

ThiefMaster commented on CHANGES.rst:
```diff
@@ -102,6 +102,7 @@ Improvements
 - Add an API for bulk-assigning contribution program codes programmatically (:issue:`5115`,
   :pr:`5120`)
 - Add layout setting to show videoconferences on the main conference page (:pr:`5124`)
+- Add forward, backward and background color option to badge designer(:pr:`5139`, thanks :user:`seginyn`)
```
should be in the 3.1 changelog

ThiefMaster commented on indico/modules/designer/client/js/index.js:
```diff
@@ -424,8 +430,62 @@ import {$T} from 'indico/utils/i18n';
+
+  function getZIndex(selectedItem, moveToFront) {
+    let zIndex = selectedItem.zIndex;
+    for (const i in items) {
```
```js
Object.values(items).forEach(item => {
  ...
});
```

and then use `item` instead of `items[i]` in there.

ThiefMaster commented on indico/modules/designer/pdf.py:
```diff
@@ -97,6 +97,8 @@ def _draw_item(self, canvas, item, tpl_data, content, margin_x, margin_y):
         style = ParagraphStyle({})
         style.alignment = ALIGNMENTS[item['text_align']]
         style.textColor = COLORS[item['color']]
+        style.backColor = item['background_color'] if item['background_color'] else None
```
I don't think we have a `background_color` for older badges, do we?

```suggestion
        style.backColor = item.get('background_color') or None
```

the ` or None` can be omitted if you do not need to normalize other falsy values such as empty strings to None

SegiNyn replied:
That right, there is no `background_color` for older badges

ThiefMaster commented on indico/modules/designer/templates/template.html:
```diff
@@ -191,7 +191,8 @@
                     <div class="subtool">
-                        <select id='color-selector' class="js-font-tool" data-attr="color">
+                        <select title="{% trans %}Text color{% endtrans %}"
+                            id='color-selector' class="js-font-tool" data-attr="color">
```
align with `title` or move into the previous line

plourenco reviewed (approved):
Added some minor suggestions but it works great, thanks! 🙌

pferreir commented on indico/modules/designer/client/js/index.js:
```diff
@@ -424,9 +430,54 @@ import {$T} from 'indico/utils/i18n';
+    let zIndex = selectedItem.zIndex;
+    Object.values(items)
+      .filter(item => selectedItem.id !== item.id)
+      .forEach(item => {
```
Not a very large performance penalty, but it would be more efficient to do the ID check inside the `forEach(...)` closure.

SegiNyn commented:
Hi, I initially used the same fixed list of colours for the background but it was suggested in an initial review that using a color picker would be preferred

pferreir commented:
@SegiNyn sure. I'm just wondering whether we shouldn't apply the same to the bg color.

ThiefMaster commented:
Consistency would clearly be useful (when making the initial suggestion I probably didn't even realized that we already have a color selector there for something)

SegiNyn commented:
Yes it would be better to keep it consistent. I'll change the colour picker to the fixed list of colours.

pferreir commented on indico/modules/designer/client/js/index.js:
```diff
@@ -424,9 +430,54 @@ import {$T} from 'indico/utils/i18n';
+          }
+        }
+      });
+    return zIndex === selectedItem.zIndex ? zIndex : zIndex + (moveToFront ? 1 : -1);
```
Untested, but maybe this would work as well?

```js
let foremostItem = _.maxBy(Object.values(items).filter(item => (selectedItem.id !== item.id) && itemsOverlap(selectedItem, item)), 'zIndex');
let zIndex = Math.min(selectedItem.zIndex, foremostItem.zIndex + 1);
```

... and I would that directly within the `back(...)` and `front(...)`, functions, using `_.maxBy(...)` and `_.minBy(...)` depending on the case. Does that make sense?

SegiNyn replied:
Yes, this works. Thank you.

pferreir commented on indico/modules/designer/client/js/index.js:
```diff
@@ -84,10 +84,12 @@ import {$T} from 'indico/utils/i18n';
       width: isImage(type) ? 150 : 400,
       height: isImage(type) ? 150 : null,
       text: $T('Fixed text'),
+      zIndex: itemIdCounter + 100,
```
Why 100 and not 10 like before?

SegiNyn replied:
This is because background images have a z-index of 5 and since moving items forward/backward adds +/-1 to the z-index value, it could easily become less than 5, placing the item behind the background image. Which is why I used 100 instead of 10.

pferreir replied:
Yeah, I get it. However, it is still possible that, after a few runs this gets again to under 5. Could we maybe have a function which "normalizes" the z-indexes after each operation? E.g. sort them in growing order, assign them z-indices according to the sequence of natural numbers (1, 2, 3, 4...) and then add 9 to all? This would keep them to a minimum. It would also hold more consistent results when you move things around.

SegiNyn replied:
Yeah, this would work

pferreir commented on indico/modules/events/registration/badges.py:
```diff
@@ -83,7 +83,8 @@ def _draw_badge(self, canvas, registration, template, tpl_data, pos_x, pos_y):
         # Print images first
         image_placeholders = {name for name, placeholder in placeholders.items() if placeholder.is_image}
-        items = sorted(tpl_data.items, key=lambda item: item['type'] not in image_placeholders)
+        items = sorted(tpl_data.items, key=lambda item: (int(item.get('zIndex', 65535)),
```
Not sure about 65535 as a default value. Do we want "legacy" items to show up in front by default? Wouldn't 0 be a better choice? And maybe a comment to explain the reasoning would be of help.

SegiNyn replied:
I don't have a reason for choosing a higher value but we can't use 0 because that would place the item behind the background image.

pferreir replied:
Yeah, well, 10, in this case.

pferreir commented:
@SegiNyn Sorry, my bad, what I meant was changing the fixed list (fg color) to a picker.

SegiNyn commented:
Okay, got it. Thanks!

pferreir commented on indico/modules/designer/client/js/index.js:
```diff
@@ -159,8 +163,8 @@ import {$T} from 'indico/utils/i18n';
     }
 
     newDiv.draggable({
+      // stack: templateSide + ' > div' remove this so it does not interfere with the z-index
```
If this is to be removed, let's just remove it. We don't usually commit commented-out code.

pferreir reviewed (changes requested):
Thanks once again. Just some minor comments/suggestions.

pferreir commented on indico/modules/designer/client/js/index.js:
```diff
@@ -424,9 +429,62 @@ import {$T} from 'indico/utils/i18n';
+    Object.values(items)
+      .sort((item1, item2) => item1.zIndex - item2.zIndex)
+      .forEach(item => {
+        item.zIndex = increament + 9;
```
"increment"

pferreir replied:
Also, this can be rewritten as (untested):

```js
_.sortBy(Object.values(items), 'zIndex').forEach((item, n) => { item.zIndex = n + 10; });
```

SegiNyn replied:
Yes, this works

pferreir commented on indico/modules/designer/client/js/index.js:
```diff
@@ -424,9 +429,62 @@ import {$T} from 'indico/utils/i18n';
+    return !(top1 >= bottom2 || top2 >= bottom1);
+  }
+
+  function normalizeZindex() {
```
I belive the right camel-case notation would be `normalizeZIndex`.

pferreir commented:
OK, thanks a lot. I think we're good code-wise, but I've found a few issues testing it:

e.g.

![image](https://user-images.githubusercontent.com/2699/145574564-f1c4b939-ba05-4ef2-a2b6-34d3eb3966d7.png)

which actually then renders OK, but the preview is weird.

Then, there is an issue with the z-index, probably side-effect of the algorithm we didn't foresee:

![Kapture 2021-12-10 at 13 34 57](https://user-images.githubusercontent.com/2699/145575155-4c434571-168f-4342-a1fc-7a147e78626a.gif)

Finally, it's currently impossible to figure out when an item is selected, if it's got a background. I suggest we move from changing the background of the item to adding a thick border around, or something like that.

pferreir commented on indico/modules/designer/client/js/index.js:
```diff
@@ -454,7 +454,7 @@ import {$T} from 'indico/utils/i18n';
           div.css('zIndex', zIndex);
           selectedItem.zIndex = zIndex;
-          normalizeZindex();
+          normalizeZIndex();
```
Maybe related to the bug we're seeing. Shouldn't we normalize before we set the CSS property?
Something like:
```js
selectedItem.zIndex = zIndex;
normalizeZIndex();
div.css('zIndex', selectedItem.zIndex);
```

pferreir commented on indico/modules/designer/client/js/index.js:
```diff
@@ -475,14 +475,10 @@ import {$T} from 'indico/utils/i18n';
-      });
+  function normalizeZIndex() {
+    _.sortBy(Object.values(items), 'zIndex').forEach((item, n) => {
+      item.zIndex = n + 10;
```
Probably related to the bug I've pointed out: `normalizeZIndex` isn't actually setting any CSS properties.

SegiNyn replied:
Yes, I think it may be because of this because moving the `normalizeZIndex` call does not fix the bug

SegiNyn commented:
Hi, "I suggest we move from changing the background of the item to adding a thick border around, or something like that." So whenever an item is selected, we add a think border around it so it is more noticeable which item was selected while still maintaining the background colour? or the background colour is removed and remains only a thick border when the item is selected?

pferreir commented:
> So whenever an item is selected, we add a thick border around it so it is more noticeable which item was selected while still maintaining the background colour?

Yes, precisely!

SegiNyn commented:
Hello @pferreir, I updated the code by adding a thicker border to selected items, make the item height use its initial value so the text is not larger than the background and updated the html zIndex values in the normalizeZIndex function

SegiNyn commented:
I added a 'data-id' attribute for the divs so they can be easily selected using:
`$('.ui-draggable[data-id=' + item.id + ']')` 
but I'm not sure if you would want this attribute. We could also use: 
 ` $('.ui-draggable').each(function(){
        if($(this).data('id') === item.id){
          $(this).css('zIndex', n + 10);
        }
     });`

SegiNyn commented:
Hello @pferreir, please could you re-review the changes (I used the github re-review button at the top but the review status was not changing). Thank you

pferreir commented:
> Hello @pferreir, please could you re-review the changes (I used the github re-review button at the top but the review status was not changing). Thank you

Hi there. Yeah, I'll have a look as soon as I can.

SegiNyn commented:
Okay, thank you!

vasantvohra commented:
Hello,
Can we merge this PR if it looks good? it would help us deploy this in our next release set for January 12th?

ThiefMaster commented:
- If I edit an existing badge (with no background color information) both colorpickers show black (which isn't correct)
- How do I clear the background color (ie revert to transparent)?

SegiNyn commented:
Hi @ThiefMaster, this was because the HTML colour picker defaults to black whenever there is no colour selected, and it does not have a transparent option

ThiefMaster commented:
I think we need to find a solution for this... at least going back to transparent background is needed for sure. Not sure what the best solution is though...  I'll ping my colleagues in case anyone has a good idea.

(it's pretty ridiculous that the native color picker doesn't provide something for this!)

SegiNyn commented:
Okay, thank you. We also saw that Indico has a `ColorPickerWidget` and I'm testing to see if this would work inplace of the native color picker.

ThiefMaster commented:
That might indeed by worth looking into. I don't think it supports clearing either but adding it there is probably easier than adding a fully new widget wrapping the native one.

SegiNyn commented:
Hello @ThiefMaster, we have updated the code to use the ColorPickerWidget, which has the option to clear the color, but this changes the look because of the extra input field
<img width="409" alt="Screenshot 2022-01-06 at 14 59 12" src="https://user-images.githubusercontent.com/54508387/148394246-bb97c71b-27e9-4569-bf38-03b661490114.png">
<img width="391" alt="Screenshot 2022-01-06 at 14 58 44" src="https://user-images.githubusercontent.com/54508387/148394290-95297a40-2e32-4ab5-b4fd-a3423971c80a.png">

ThiefMaster commented:
Fine for me, I'd just apply some CSS so the whole color code fits in the text field and the color box has the same size as the field (and is maybe attached right to its left)

SegiNyn commented:
It now looks like this:
<img width="431" alt="Screenshot 2022-01-07 at 16 58 09" src="https://user-images.githubusercontent.com/54508387/148572086-8662aa15-43cd-4cca-b48b-f1e8548ffa0a.png">

ThiefMaster commented:
Much nicer. Could you "equalize" the heights and remove the spacing between the color display and the text field?

SegiNyn commented:
<img width="385" alt="Screenshot 2022-01-07 at 17 36 21" src="https://user-images.githubusercontent.com/54508387/148575857-87b802d6-7742-4444-933f-ce03205c0029.png">

ThiefMaster commented:
almost :) there are still rounded top-left/bottom-left corners on the text fields, but since that's no longer an outer border these corners should be regular now

SegiNyn commented:
<img width="383" alt="Screenshot 2022-01-07 at 18 05 29" src="https://user-images.githubusercontent.com/54508387/148580080-d3ed402d-8f2b-42fc-8ac5-88e919f179c6.png">

ThiefMaster commented on CHANGES.rst:
```diff
@@ -46,6 +46,7 @@ Improvements
   settings (:issue:`4837`, :issue:`2224`, :issue:`1877`, :pr:`5179`)
 - Allow filtering the contribution list based on whether any person (speaker or author)
   has registered for the event or not (:issue:`5192`, :pr:`5193`)
+- Add forward, backward and background color option to badge designer(:pr:`5139`, thanks :user:`seginyn`)
```
```suggestion
- Add background color option and layer order to badge/poster designer items (:pr:`5139`,
  thanks :user:`SegiNyn`)
```

ThiefMaster commented on indico/modules/designer/client/js/index.js:
```diff
@@ -240,11 +244,28 @@ import {$T} from 'indico/utils/i18n';
+    }
+    $('#bg-color-selector').val(item.background_color);
+    var bgColorPreview = $('#bg-color-preview');
+    if (bgColorPreview.hasClass('no-value') && item.background_color.length > 1) {
```
`item.background_color` is null for existing badges. Maybe just check `item.background_color` which will cover both empty strings and null/undefined?

Also, why `> 1` and not `> 0` in this check?

ThiefMaster commented on indico/modules/designer/client/js/index.js:
```diff
@@ -240,11 +244,28 @@ import {$T} from 'indico/utils/i18n';
     // Change the selectors so that they match the properties of the item
+    $('#text-color-selector').val(item.color);
+    var textColorPreview = $('#text-color-preview');
+    if (textColorPreview.hasClass('no-value') && item.color.length > 1) {
```
No need for that hasClass check; removing a class that isn't there is just a no-op. And I think you can greatly simplify it like this:

```
textColorPreview.toggleClass('no-value', !item.color);
```

ThiefMaster commented on indico/modules/designer/client/js/index.js:
```diff
@@ -240,11 +244,28 @@ import {$T} from 'indico/utils/i18n';
+    if (textColorPreview.hasClass('no-value') && item.color.length > 1) {
+      textColorPreview.removeClass('no-value');
+    }
+    textColorPreview.attr('style', 'background-color: ' + item.color + '; opacity: 1;');
```
You can use `.css({...})` to set the style directly instead of using a css string

ThiefMaster commented on indico/modules/designer/client/js/index.js:
```diff
@@ -240,11 +244,28 @@ import {$T} from 'indico/utils/i18n';
 
     // Change the selectors so that they match the properties of the item
+    $('#text-color-selector').val(item.color);
+    var textColorPreview = $('#text-color-preview');
```
Maybe we could use `const`/`let` in new code, even if the rest of the file is very old and still uses `var`?

ThiefMaster commented on indico/modules/designer/templates/template.html:
```diff
@@ -191,18 +191,24 @@
+                            <div class="clickable-wrapper">
+                                <span id="text-color-preview" class="color-preview"></span>
+                                <input title="{% trans %}Text color{% endtrans %}" type="text"
+                                    id="text-color-selector" value="#" class="js-font-tool" data-attr="color"
+                                    maxlength="7" size="2">
```
alignment

ThiefMaster commented on indico/modules/designer/templates/template.html:
```diff
@@ -288,6 +294,14 @@
                                 class="i-button move-button icon-only text-color subtle icon-expand"
                                 data-direction="bottom">
                         </button>
+                        <button title="{% trans %}Send item to the back{% endtrans %}"
```
Let's be consistent on the titles, e.g.:

- Move item to the back
- Move item to the front
